### PR TITLE
[4.0] Add display field attibute

### PR DESCRIPTION
**Pull request by ciar4n**
Pull Request for Issue  #32487 .

### Summary of Changes
Because the 'class'  XML attribute adds the class to the field itself rather than the outside container, it is not suitable for the classes added to  #32422. This PR adds a `parentclass` attribute which adds a class to the outside `control-group` div.

### Testing Instructions
Ensure https://github.com/joomla/joomla-cms/pull/32422 works as described (updated to match)

### CODE REVIEW

**Comment by drmenzelit** [comment]
I have tested this item :white_check_mark: successfully on 7cef8cf9d9e44858f18addad012a4f567dac892c<br /><br />Tested in templateDetails.xml from Cassiopeia.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32488">issues.joomla.org/tracker/joomla-cms/32488</a>.</sub>

File: layouts/joomla/form/renderfield.php [hunk 1]
```diff
@@ -35,6 +35,11 @@
 $id              = $name . '-desc';
 $hide            = empty($options['hiddenLabel']) ? '' : ' visually-hidden';
 $hideDescription = empty($options['hiddenDescription']) ? false : $options['hiddenDescription'];
+
+if($parentclass) {
+	$class .= ' ' . $parentclass;
+}
```
Comment (Quy) [inline layouts/joomla/form/renderfield.php]:
```suggestion
if ($parentclass)
{
	$class .= ' ' . $parentclass;
}
```

File: libraries/src/Form/Field/CalendarField.php [hunk 2]
```diff
@@ -81,6 +81,13 @@ class CalendarField extends FormField
 	 */
 	protected $layout = 'joomla.form.field.calendar';
 
+	/**
+	 * The parent class of the field
+	 *
+	 * @var  string
+	 */
```
Comment (Quy) [inline libraries/src/Form/Field/CalendarField.php:89]: Please add `@since  __DEPLOY_VERSION__`.

Comment (ciar4n) [reply libraries/src/Form/Field/CalendarField.php:89]: Thank you @Quy 👍

Comment (Quy) [reply libraries/src/Form/Field/CalendarField.php:89]: This has to be done with the other instances. 😞 Thanks.

Comment (ciar4n) [reply libraries/src/Form/Field/CalendarField.php:89]: @Quy .. all done!

**Comment by chmst** [comment]
I have tested in com_content, article.xml and there it works fine.

grafik

But in Options (config.xml) there is no effect. 

grafik

**Comment by ciar4n** [comment]
@chmst Thank you for testing. 

Strangely I am unable to replicate this issue. Adding to ` ..\administrator\components\com_config\forms\application.xml` for example `parentclass="span-2"` works without issue.

image

**Comment by ciar4n** [comment]
@chmst My apologies. I see the issue now in component tabs as you said. I'll submit a fix.

**Comment by ciar4n** [comment]
@chmst Should be good now. 👍

**Comment by chmst** [comment]
I have tested this item :white_check_mark: successfully on 780fdc49853d2ea885bdb8e3834a848b3a777de8<br /><br />Tested wit com_content: article edit, com_content config.xml, com_users: user edit
and an own J4 component; 
Documentation will be required. It is a great improvement. Thank you. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32488">issues.joomla.org/tracker/joomla-cms/32488</a>.</sub>

**Comment by drmenzelit** [comment]
I have tested this item :white_check_mark: successfully on 780fdc49853d2ea885bdb8e3834a848b3a777de8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32488">issues.joomla.org/tracker/joomla-cms/32488</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32488">issues.joomla.org/tracker/joomla-cms/32488</a>.</sub>

**Comment by korenevskiy** [comment]
1. Using the **`display`** attribute is confusing, because after the **`display`** you want to write **`block`**. Yes, in general, and **`parentclass`** is also not clear about which parent we are talking about. I suggest using the **`controlclass`** attribute, we know that in the full sense of the control is just **`<div class= "control-group"/>`**
2. I suggest making the layout multi-column. But with the observance of several rules. the **`note`** field and the **`spacer`** field will be full-width.
Controls located between the **`note`** and **`spacer`** fields, only between them will be distributed in 2-3 columns. This will allow you not to create a garbage can from the fields. And of course the ability to use the **`controlclass`** attribute.
But having the ability to override the default classes and set your own width in the number of columns.
.
In any case, now in Joomla Nightly 1 column mode, by analogy J3,  looks better. 1 column is clearer and the most important thing is available. I propose to center 1 column in the middle.
.
How do I output a control-table without a Label?
img-2021-03-07-23-58-39

**Comment by brianteeman** [comment]
`hiddenLabel="true"`

**Comment by infograf768** [comment]
@ciar4n 

just a note: the code proposed is
```
if ($parentclass)
{
	$class .= ' ' . $parentclass;
}
```

As it already adds a space, we do not need the space in your example in https://github.com/joomla/joomla-cms/pull/32422
```
<field
	name="article_layout"
	type="componentlayout"
	label="JFIELD_ALT_LAYOUT_LABEL"
	class="form-select"
	useglobal="true"
	extension="com_content"
	view="article"
        parentclass=" span-4 stack"
/>
```

also, I do not understand why, in this PR you add stack to the article layout in  `administrator/components/com_content/forms/article.xml` as it does not have any effect.

https://github.com/joomla/joomla-cms/pull/32488/files#diff-48dcb2f2108f7b07952c9c407fd23cab28677e3d2fd437030b2f04a043282320R280-R286

**Comment by ciar4n** [comment]
@infograf768 My bad, I didn't mean to commit https://github.com/joomla/joomla-cms/pull/32488/files#diff-48dcb2f2108f7b07952c9c407fd23cab28677e3d2fd437030b2f04a043282320R280-R286 (reverted with https://github.com/joomla/joomla-cms/pull/32488/commits/cb121a57994836a02a7355b6c2dc1d632a9a6d80). The PR can remain in RTC as this class was ineffective.

And yes, no space is required in the example on #32422. I have edited this accordingly. Thanks for mentioning it!

**Comment by ciar4n** [comment]
@korenevskiy I guess the logic of `parentclass` is that this class is been added to the parent div of the field that the rule is creating. Unlike the `class` attribute, this does not add it to the `control` class div.

**Comment by korenevskiy** [comment]
> @korenevskiy I guess the logic of `parentclass` is that this class is been added to the parent div of the field that the rule is creating. Unlike the `class` attribute, this does not add it to the `control` class div.

The control is `< div class= "control-group" >`. But the field is `< input >`. The field is located inside the control. The field has a parent that is a control. But the control also has a parent. It is intuitive for a layout designer who is unfamiliar with Joomla, but knows Bootstrap. ask which parent is being referred to when the class is specified, the parents of the control, or the parents of the field? But by specifying controlclass, it is clear that this is the class of the tag `< div class="control group" >` .

let the class be added directly to `<div class= "control-group">`
this will allow you to position the `label` at the same time.
Or can use another `fieldclass `class

**Comment by korenevskiy** [comment]
by the way, after all these tricks, the next step is this. 1. Creating an XML element to connect the CSS file to the form. 2. Create an XML tag to include Styles in the form.. Maybe this will be the right decision. This will be bad form for developers, but supporting such a feature is the right solution. I've seen modules where CSS is plugged in inside a custom field. This is worse than connecting directly to XML. but Joomla does not know how to do this.

You can consider a new attribute for `<field>`, such as `<field type="text" css="my_module.css"/>`

**Comment by korenevskiy** [comment]
> @korenevskiy I guess the logic of `parentclass` is that this class is been added to the parent div of the field that the rule is creating. Unlike the `class` attribute, this does not add it to the `control` class div.

img-2021-03-09-13-35-32
it is not clear from the name of the attribute which parent we are talking about
Not intuitive and unclear.
.
But **`<field controlclass="span-4 stack">`** -> **`<div class= "control-group span-4 stack">`**
And now control will have a width across the entire GRID.

**Comment by ciar4n** [comment]
@korenevskiy As this PR is already set to RTC, I'll be unlikely to edit it further. Feel free to create a PR against this code once/if this gets merged 👍

**Pull request merged**